crates/sui-framework/src/lib.rs

```diff
@@ -132,6 +137,58 @@ pub fn run_move_unit_tests(
     )
 }
 
+pub fn build_move_package(
+    path: &Path,
+    build_config: BuildConfig,
+) -> SuiResult<Vec<CompiledModule>> {
+    let pkg = build_move_package_with_deps(path, build_config)?;
+    verify_framework_version(&pkg)?;
+    filter_package_modules(&pkg)
+}
+
+/// Version of the framework code that the binary used for compilation expects should be the same as
+/// version of the framework code bundled as compiled package's dependency and this function
+/// verifies this.
+fn verify_framework_version(pkg: &CompiledPackage) -> SuiResult<()> {
+    // We stash compiled modules in the Modules map which is sorted so that we can compare sets of
+    // compiled modules directly.
+
+    let dep_framework_modules = pkg.all_modules_map().iter_modules_owned();
+    let dep_framework: Vec<&CompiledModule> = dep_framework_modules
+        .iter()
+        .filter(|m| *m.self_id().address() == SUI_FRAMEWORK_ADDRESS)
+        .collect();
+
+    let framework_modules = Modules::new(get_sui_framework().iter()).iter_modules_owned();
+    let framework: Vec<&CompiledModule> = framework_modules.iter().collect();
+
+    if dep_framework != framework {
+        return Err(SuiError::ModuleVerificationFailure {
+            error: "Sui framework version mismatch between package dependencies and Sui binary"
```

Comment:
We should give a suggestion of what to do here. What would the standard thing be to do? Just say "make sure your `sui` command line tool was built from the same branch/git checkout you are using for your Sui Framework files" (something like that)